### PR TITLE
[WIP] Add support for Markdown in tweets

### LINKED ISSUE #39: Syntax highlighting should remove links inside code blocks
See: https://twitter.com/tomasdev/status/953720238784303105

// @tomma5o

### LINKED ISSUE #41: Allow a subset of Markdown
With code block support, I think it would make sense to allow a real markdown (subset).
The most common things I would use (and you can sometimes see in use on Twitter) are bold, italic, lists and inline code.

The "document"-oriented markup like headings could get abused and don't really make sense in a tweet, as well as links and images, which Twitter already handles differently.

### LINKED ISSUE #108: Code block ruins link
See this tweet: https://twitter.com/sindresorhus/status/994307954030526464

The `https://github.com/sindresorhus/sindresorhus` URL is a link when Refined Twitter is turned off. When it's on, it's just text. For some reason it's also showing the URL to the image as text `pic.twitter.com/xZdQSzFLAq`.

// @jorgegonzalez

### LINKED ISSUE #110: Remove cards from links from code blocks
The links rendered from code blocks should be detected and removed from tweet cards. 

Basically turn this:
image

Into this:
image

### LINKED ISSUE #123: Inline code blocks not applied correctly
See this tweet: https://twitter.com/MartijnSaly/status/1011223724496957441

image

Couldn't find the reason in the extension's source code, fiddling with it in the dev tools does everything as expected...

### PR DESCRIPTION
**Pull request by jorgegonzalez**
Fixes #39 
Fixes #41 
Fixes #108 
Fixes #110 
Closes #109 
Fixes #123 

@filipekiss I'm sorry that this duplicates all of the work you've been doing in #109! The current solution we had for the inline code and the highlighted code blocks was pretty messy. This PR should finally fix the issues with links in the styled code. It also adds support for all other markdown.

It should also be noted that `remark-html`, `remark-parse`, and `unified` are adding about 100KB to the bundle, so this may not be worth it, though the packaged chrome extension will only go from 50KB to 92KB, which isn't _too_ crazy.

Todo:

- [x] Fix highlighted code padding
- [x] Fix header padding
- [x] Style blockquotes
- [x] Style lists
- [x] Possibly tweak `strong` and `em` elements
- [x] Possibly strip `http://` from links in inline code
- [ ] Fix styles in tweet modals
	- https://twitter.com/rauschma/status/998556286538133505
- [ ] Remove link cards if they originate from code blocks
~~Use smart quotes https://twitter.com/zats/status/996730907192102912~~

Issues:
- [x] Some links in the stream are getting squashed 
	- Ignoring as I couldn't reproduce
	- Was related to the issue below, now fixed
- [x] Space before <a> tags in non-processed tweets
	- Fixed by ignoring link tags in non-markdown tweets
- [x] Code blocks are not styled when language isn't specified
	- https://twitter.com/sindresorhus/status/994307954030526464

https://twitter.com/mikeal/status/1011406440999763968

~~Won't work on this PR anymore until the bundle size issue is discussed.~~

### CODE REVIEW

**Comment by jorgegonzalez** [comment]
image

**Comment by jorgegonzalez** [comment]
image

**Comment by filipekiss** [comment]
@jorgegonzalez No problem at all! This is a great solution and also adds Markdown support, which I'm 100% ok with!

**Review by filipekiss — CHANGES_REQUESTED** [review]

**Comment by filipekiss** [comment]
About the bundle size: I think it's negligible. The value added is enormous and the extension itself isn't big (although it almost doubles size).

The extension will be installed and then done. There's no need to worry about size so much because there are no network requests or anything like that after the extension is installed/updated, so I'm perfectly fine with it.

**Comment by jorgegonzalez** [comment]
We are currently formatting the smart quotes style (and the little guys hands), anyone know why this is? twitter.com/zats/status/996730907192102912

image

image

**Comment by filipekiss** [comment]
We're not actively doing this. Probably when we parse the tweet to add markdown support (or something), using `Element.textContent` get's rid of the smart quotes.

**Comment by jorgegonzalez** [comment]
The master branch does it too, and when inspecting the `<p>` element, the smart quotes are still there.

Not a major issue though, I think.

**Comment by sindresorhus** [comment]
This is amazing! I'm ok with the size increase.

**Comment by jorgegonzalez** [comment]
Awesome, I'll work out the kinks soon.

**Comment by jorgegonzalez** [comment]
Possible issue https://twitter.com/rauschma/status/998556286538133505

``'a|b|c'.split`|` ``

**Comment by filipekiss** [comment]
Not sure if it's actually an issue. That's not standard, it's just abusing the syntax/language to show how stuff works. I wouldn't mind much.

**Comment by jorgegonzalez** [comment]
Yeah I don't mind either, just trying to document encounters like this.

**Comment by jorgegonzalez** [comment]
@'s are not being tokenized properly.

**Comment by jorgegonzalez** [comment]
> this is a test

**Comment by jorgegonzalez** [comment]
https://twitter.com/DiscreteQuantum/status/1005474353151127552

image

**Comment by jorgegonzalez** [comment]
I think this PR is close to finished. I need to tweak a few more things, but I'd appreciate it if anyone wants to start testing the branch! Also any feedback.

**Comment by jorgegonzalez** [comment]
Multiline code blocks are getting smushed. CSS is hard.

**Comment by jorgegonzalez** [comment]
https://twitter.com/styfle/status/1008762725282000896

File: source/features/markdown.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+		'</del>'
+	].filter(tag => html.includes(tag)).length > 0;
+
+function stripLinks(el) {
+	const possibleLink = $(el).find('span.token.string');
+
+	possibleLink.each((i, el) => {
+		const {innerText} = el;
+
+		if (innerText.includes('http://')) {
```
Comment (sindresorhus) [inline source/features/markdown.js]: Use `textContent` over `innerText`. Applies to all the other cases too.

File: source/features/markdown.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		// Ensure this is only run once
+		if ($(el).hasClass(processed)) {
+			return;
+		}
+
+		unified()
+			.use(markdown)
+			.use(html)
+			.use(tokenizer)
+			.process(el.textContent, (err, file) => {
```
Comment (sindresorhus) [inline source/features/markdown.js]: If you don't pass the callback it will return a promise instead. Async/await FTW.

https://github.com/unifiedjs/unified#processorrunnode-file-done

File: source/libs/tokenizer.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
+/**
+ * This custom tokenizer for remark-parse is needed so that
+ * hashtags (#refinedtwitter) and mentions (@sindresorhus) are
+ * not sanitized when the markdown is compiled to HTML.
+ */
+
+function mentions(eat, value, silent) {
+	const match = /^@(\w+)/.exec(value);
+
+	if (match) {
```
Comment (sindresorhus) [inline source/libs/tokenizer.js:10]: Can you destructure `match` here and give the parts names to make it more readable?

File: source/libs/tokenizer.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		if (silent) {
+			return true;
+		}
+
+		return eat(match[0])({
+			type: 'link',
+			url: 'https://twitter.com/' + match[1],
```
Comment (sindresorhus) [inline source/libs/tokenizer.js]: Use template literal

File: source/style/content.css [hunk 5]
```diff
@@ -41,11 +43,13 @@ body {
 }
 
 /* Highlight your mentions in the stream */
-body:not(.NotificationsPage) .tweet.refined-twitter_mention:not(.permalink-tweet) {
+body:not(.NotificationsPage)
+	.tweet.refined-twitter_mention:not(.permalink-tweet) {
```
Comment (sindresorhus) [inline source/style/content.css]: Don't do unrelated changes.

File: package.json [hunk 6]
```diff
@@ -1,4 +1,8 @@
 {
+	"name": "refined-twitter",
+	"private": true,
+	"repository": "https://github.com/sindresorhus/refined-twitter",
+	"license": "MIT",
```
Comment (sindresorhus) [inline package.json]: Why add this? It's not needed since it's not going to be published on npm.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: package.json (hunk 6 above)
Comment (jorgegonzalez) [reply package.json]: npm was complaining

```
npm WARN refined-twitter No repository field.
npm WARN refined-twitter No license field.
```

Comment (sindresorhus) [reply package.json]: Just ignore npm on that. They don't require it at all so not sure why they would warn on it.

File: source/libs/tokenizer.js (hunk 3 above)
Comment (jorgegonzalez) [reply source/libs/tokenizer.js:10]: Not sure what you mean

File: source/style/content.css (hunk 5 above)
Comment (jorgegonzalez) [reply source/style/content.css]: My editor auto-formats everything through Prettier.

File: source/libs/tokenizer.js (hunk 3 above)
Comment (sindresorhus) [reply source/libs/tokenizer.js]:
```jsif (match) {

Comment (sindresorhus) [reply source/libs/tokenizer.js]:
```js

Comment (sindresorhus) [reply source/libs/tokenizer.js:10]:
```js
if (match) {
  const [foo, bar] = match;
```

With `foo` and `bar` having relevant names.

File: package.json [hunk 7]
```diff
@@ -34,6 +37,10 @@
 		"webpack-cli": "^2.0.10",
 		"xo": "*"
 	},
+	"prettier": {
+		"bracketSpacing": false,
+		"singleQuote": true
+	},
```
Comment (sindresorhus) [inline package.json]: This doesn't belong here. The project doesn't use Prettier. You should rather just locally disable Prettier for this project.

File: source/features/markdown.js [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+		.html(`<div>${highlightedCode}</div>`);
+
+	stripLinks(compiledEl);
+}
+
+export default () => {
+	const processed = 'refined-twitter_markdown-processed';
+	const styledClassName = 'refined-twitter_markdown';
+
+	$('.tweet-text').each(async (i, el) => {
```
Comment (sindresorhus) [inline source/features/markdown.js]: I have a feeling the markdown compiling is pretty slow, so would be good to use `Promise.all()` here to run it concurrently.

File: source/libs/tokenizer.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+mentions.notInLink = true;
+mentions.locator = (value, fromIndex) =>
+	value.indexOf('@', fromIndex);
+
+function hashTags(eat, value, silent) {
+	const match = /^#(\w+)/.exec(value);
+
+	if (match) {
```
Comment (sindresorhus) [inline source/libs/tokenizer.js:31]: Same as https://github.com/sindresorhus/refined-twitter/pull/113/files#r197670614

**Comment by jorgegonzalez** [comment]
Emojis are being removed https://twitter.com/MartijnSaly/status/1011223724496957441

File: package.json (hunk 7 above)
Comment (jorgegonzalez) [reply package.json]: Disabling Prettier in my editor would be very counterproductive for me. This was just to keep the formatting consistent with XO and it doesn't make a difference to anything else.

File: source/libs/tokenizer.js (hunk 3 above)
Comment (jorgegonzalez) [reply source/libs/tokenizer.js:10]: I don't think this change will add much. `match[0]` is always something like `#sindresorhus` or `@sindresorhus` and `match[1]` is always the string with the `#` or `@` removed (`sindresorhus`).

Comment (jorgegonzalez) [reply source/libs/tokenizer.js:10]: ~~Well I at least can't think of good names for these.~~

**Comment by jorgegonzalez** [comment]
Sorry this is taking so long. In the time since I opened this PR I graduated university, went on holiday, moved to a new city, and started a new job. Will try to work on it more this weekend!

**Comment by jorgegonzalez** [comment]
So emojis inside of inline code blocks (`like this 🔥`) are being deleted. I can't really figure out why.

I feel like that's not a common use case. Should we fix that issue in a different PR?

**Comment by filipekiss** [comment]
@jorgegonzalez I'm okay with it being deleted for the time being and we work on that later.

**Comment by notlmn** [comment]
Emojis are converted to `` tags by Twitter, so you're almost screwed over there.

When you call `element.textContent`, the result would be

```
Okay okay, then how about calling `` an error and `` an event. Hey javascript allows it!
                                   ^-- emoji gone from here
```

You'd need to specially handle DOM nodes of the tweet, and serialize them!

**Comment by jorgegonzalez** [comment]
Might be able to write a tokenizer for this.

**Comment by jorgegonzalez** [comment]
https://twitter.com/swyx/status/1023762646758240256

File: readme.md [hunk 10]
```diff
@@ -102,14 +109,15 @@ Once built, load it in the browser of your choice:
 
 ## FAQ
 
-### Don't you have another extension with the same name?
+#### Don't you have another extension with the same name?
```
Comment (notlmn) [inline readme.md]: Why is this 4?

**Comment by superhawk610** [comment]
I have another failing example, it may be fixed by the existing changes in this PR, not sure.

screen shot 2019-02-07 at 2 43 43 pm

Original text:

Reading up on `Object.is` vs `===` - is there any world where I would care to differentiate between `+0` and `-0`? What about `NaN` and `Number.NaN`? Why?

@jorgegonzalez @filipekiss what's the current status on this PR? I may be able to help wrap it up, just want to confirm that no one else is actively working on it.

**Comment by jorgegonzalez** [comment]
I had been actively working on it but stopped due to various issues I kept running into, and lack of time as I started a full time job. The branch is open and I’ll merge anything you want to contribute.

**Comment by superhawk610** [comment]
Alright, I'll track my progress here:

- [ ] retweet icon is missing in Timeline and Tweet modal
- [ ] quotes only take up a few pixels in height
- [ ] extension errors on load in Chrome
```
Uncaught EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self' blob: filesystem: chrome-extension-resource:".
```

**Comment by jorgegonzalez** [comment]
@superhawk610 any updates?

I've had less and less time as the year has gone on to work on this.

**Comment by superhawk610** [comment]
After an initial bit of work, working on this keeps losing out to interview prepping 😬 I would still like to finish this, but if anyone else is able to contribute sooner than me then that would also be great.

I'll try to put some time towards it this weekend.

**Comment by jorgegonzalez** [comment]
> working on this keeps losing out to interview prepping 

I'm right there with ya, bud.

**Comment by sindresorhus** [comment]
I'm closing this for now as this probably broke with the new Twitter redesign and it hasn't had any activity for some time.
